### DISCUSSION TITLE
Nullable Support

### DISCUSSION THREAD

**Discussion by AmalDarKhalil** [riok/mapperly#578] [Q&A] [answered]
If we had an `EmployeePosition` and it has `PayRate?` which could be null, the mapper will be broken if the Payrate is null, any idea of what I can do to fix it?
![image](https://github.com/riok/mapperly/assets/85879334/0b970720-64c3-4798-884a-a35ca3197cde)

**Comment by TimothyMakkison** [comment, marked as answer]
Hey, thanks for using mapperly.

Do you have a simple example you could share or possibly the generated code? Mapperly should detect nullable members and generate conditional mapping code.
If you've recently added mapperly then sometimes restarting your ide can help.

### Example nullable:
```C#

[Mapper]
public static partial class Mapper
{
    public static partial EmployeePositionDto Map(EmployeePosition car);
}

public class EmployeePosition { public PayRate? PayRate { get; set; } }

public class EmployeePositionDto { public PayRateDto PayRate { get; set; } }

public class PayRate {}

public class PayRateDto {}
```
### Generated code 
```C#
public static partial global::Riok.Mapperly.Sample.EmployeePositionDto Map(global::Riok.Mapperly.Sample.EmployeePosition car)
{
    var target = new global::Riok.Mapperly.Sample.EmployeePositionDto();
    if (car.PayRate != null)
    {
        target.PayRate = MapToPayRateDto(car.PayRate);
    }

    return target;
}

private static global::Riok.Mapperly.Sample.PayRateDto MapToPayRateDto(global::Riok.Mapperly.Sample.PayRate source)
{
    var target = new global::Riok.Mapperly.Sample.PayRateDto();
    return target;
}
```

**Reply by AmalDarKhalil** [reply]
oh, so there's no need to create PayRate mapper?
and thanks!

**Reply by AmalDarKhalil** [reply]
can you please tell me where I can find the generated code? I'm using rider

**Reply by TimothyMakkison** [reply]
> oh, so there's no need to create PayRate mapper?

Mapperly should generate mappers for all members, [see example here](https://mapperly.riok.app/docs/getting-started/generated-mapper-example/). If you need more control over the generated mappers then Mapperly will try to use your other defined mappers.

> can you please tell me where I can find the generated code? I'm using rider

~~Iirc with rider you can select the method or mapper and hit `f7`. [See also](https://mapperly.riok.app/docs/configuration/generated-source)~~

**Reply by TimothyMakkison** [reply]
> can you please tell me where I can find the generated code? I'm using rider

My bad, in rider select the method or mapper and press `F12`. [See also](https://mapperly.riok.app/docs/configuration/generated-source)

**Reply by AmalDarKhalil** [reply]
thanks for your support! I checked the generated code and it turned out we have a cycle in the models
like e.g. Employee has Position and the Position has Employee.

**Reply by TimothyMakkison** [reply]
> thanks for your support! 

Happy to help!

>I checked the generated code and it turned out we have a cycle in the models
like e.g. Employee has Position and the Position has Employee.

Mapperly should be able to handle that, check out [reference handling](https://mapperly.riok.app/docs/configuration/reference-handling)

```C#
[Mapper(UseReferenceHandling = true)]
public partial class EmployeeMapper
{
    public partial EmployeeDto MapEmployee(Employee src);
}
```
